src/rulesets/Base/indexing.jl

```diff
@@ -54,38 +54,77 @@ end
 
 
 #####
-##### getindex
+##### getindex(::AbstractArray)
 #####
 
 function frule((_, ẋ), ::typeof(getindex), x::AbstractArray, inds...)
     return x[inds...], ẋ[inds...]
 end
 
-function rrule(::typeof(getindex), x::Array{<:Number}, inds...)
+function rrule(::typeof(getindex), x::AbstractArray, inds...)
     # removes any logical indexing, CartesianIndex etc
     # leaving us just with a tuple of Int, Arrays of Int and Ranges of Int
     plain_inds = Base.to_indices(x, inds)
     y = getindex(x, plain_inds...)
     function getindex_pullback(ȳ)
-        function getindex_add!(Δ)
-            # this a optimizes away for simple cases
-            for (ȳ_ii, ii) in zip(ȳ, Iterators.product(plain_inds...))
-                Δ[ii...] += ȳ_ii
-            end
-            return Δ
-        end
-
-        x̄ = InplaceableThunk(
-            getindex_add!,
-            @thunk(getindex_add!(zero(x))),
+        xthunk = InplaceableThunk(
+            x̄ -> ∇getindex!(x̄, x, unthunk(ȳ), plain_inds...),
+            @thunk(∇getindex(x, unthunk(ȳ), plain_inds...)),
         )
```

Comment:
Possibly this `InplaceableThunk` should be returned by `∇getindex` immediately? I didn't because the rule for `unique` calls `reshape(∇getindex(vec(x), vec(dy), keep), axes_x)` but I have forgotten why.